findcoordinator.go

```diff
@@ -2,8 +2,120 @@ package kafka
 
 import (
 	"bufio"
+	"context"
+	"fmt"
+	"net"
+	"strings"
+	"time"
+
+	"github.com/segmentio/kafka-go/protocol/findcoordinator"
+)
+
+// CoordinatorKeyType is used to specify the type of coordinator to look for
+type CoordinatorKeyType int8
+
+const (
+	// CoordinatorKeyTypeConsumer type is used when looking for a Group coordinator
+	CoordinatorKeyTypeConsumer CoordinatorKeyType = 0
+
+	// CoordinatorKeyTypeTransaction type is used when looking for a Transaction coordinator
+	CoordinatorKeyTypeTransaction CoordinatorKeyType = 1
 )
 
+// FindCoordinatorRequest is the request structure for the FindCoordinator function
+type FindCoordinatorRequest struct {
+	// Address of the kafka broker to send the request to.
+	Addr net.Addr
+
+	// The coordinator key.
+	Key string
+
+	// The coordinator key type. (Group, transaction, etc.)
+	KeyType CoordinatorKeyType
+}
+
+// FindCoordinatorResponseCoordinator contains details about the found coordinator
+type FindCoordinatorResponseCoordinator struct {
+	// NodeID holds the broker id.
+	NodeID int32
+
+	// Host of the broker
+	Host string
+
+	// Port on which broker accepts requests
+	Port int32
+}
+
+// FindCoordinatorResponse is the response structure for the FindCoordinator function
+type FindCoordinatorResponse struct {
+	// The Transaction/Group Coordinator details
+	Coordinator *FindCoordinatorResponseCoordinator
+
+	// The amount of time that the broker throttled the request.
+	Throttle time.Duration
+
+	// An error that may have occurred while attempting to retrieve Coordinator
+	//
+	// The error contains both the kafka error code, and an error message
+	// returned by the kafka broker.
+	Error error
+
+	// Error code
+	ErrorCode int
+}
+
+// FindCoordinator sends a findCoordinator request to a kafka broker and returns the
+// response.
+func (c *Client) FindCoordinator(ctx context.Context, req *FindCoordinatorRequest) (*FindCoordinatorResponse, error) {
+
+	m, err := c.roundTrip(ctx, req.Addr, &findcoordinator.Request{
+		Key:     req.Key,
+		KeyType: int8(req.KeyType),
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("kafka.(*Client).FindCoordinator: %w", err)
+	}
+
+	res := m.(*findcoordinator.Response)
+	coordinator := &FindCoordinatorResponseCoordinator{
+		NodeID: res.NodeID,
+		Host:   res.Host,
+		Port:   res.Port,
+	}
+	ret := &FindCoordinatorResponse{
+		Throttle:    makeDuration(res.ThrottleTimeMs),
+		Error:       makeError(res.ErrorCode, res.ErrorMessage),
+		ErrorCode:   int(res.ErrorCode),
+		Coordinator: coordinator,
+	}
+
+	return ret, ret.Error
+}
+
+// WaitForCoordinatorIndefinitely is a blocking call till a coordinator is found
+func (c *Client) WaitForCoordinatorIndefinitely(ctx context.Context, req *FindCoordinatorRequest) (*FindCoordinatorResponse, error) {
```

Comment:
I would prefer to not include this method, and keep the `kafka.(*Client)` API a 1:1 mapping to Kafka APIs.

The one-second blocking call may also not be the right fit for many applications, which could need exponential backoff, or respecting context cancellation, etc...

It isn't very complex for a program to re-implement this logic, maybe we could even put it as an example in the `kafka.(*Client).FindCoordinator` documentation?

Reply:
Makes sense. This shouldn't be a part of the `kafka.(*Client)` API. Will do the needful.